### PR TITLE
Adding a directory for QA 

### PR DESCRIPTION
**Pull request by VestigeJ**
The intent is for this directory to contain tooling for QA frequently used in our day to day on nodes under test. The manifests are some of our more frequently used test workloads, in the future we hope to add some quick tests to run against a cluster in this repo adjacent to these manifests.

### CODE REVIEW

**Comment by VestigeJ** [comment]
Working through shellcheck analysis so feel free to delay reviews for now.

File: qa/qa-captain/bashrc [hunk 1]
```diff
@@ -0,0 +1,1116 @@
+#! /bin/bash
```
Comment (briandowns) [inline qa/qa-captain/bashrc:1]: Are you using Bash specific features in this? If not, change this to `/bin/sh` as not all systems have bash and not all systems have bash at that location.

File: qa/qa-captain/bashrc [hunk 2]
```diff
@@ -0,0 +1,1116 @@
...
+# Add an "alert" alias for long running commands.  Use like so:
+#   sleep 10; alert
+alias alert='notify-send --urgency=low -i "$([ $? = 0 ] && echo terminal || echo error)" "$(history|tail -n1|sed -e '\''s/^\s*[0-9]\+\s*//;s/[;&|]\s*alert$//'\'')"'
+
+# Alias definitions.
+# You may want to put all your additions into a separate file like
+# ~/.bash_aliases, instead of adding them here directly.
+# See /usr/share/doc/bash-doc/examples in the bash-doc package.
+# shellcheck source=/dev/null
+if [ -f /home/"$USER"/.bash_aliases ]; then
```
Comment (briandowns) [inline qa/qa-captain/bashrc:103]: Could we use `${HOME}` here instead?

File: qa/qa-captain/bashrc [hunk 3]
```diff
@@ -0,0 +1,1116 @@
...
+# --- prints out node token ---
+function get.token() {
+    local PRODUCT="${1:-k3s}"
+    sudo cat /var/lib/rancher/"$PRODUCT"/server/node-token
+}
+
+# --- killall uninstall all ---
+function void() {
+    # sudo systemctl list-units | grep -i -e k3s -e rke2 ??? find whats installed - then just remove those instead of trying all
+    mkdir ~/tmp-confs/
```
Comment (briandowns) [inline qa/qa-captain/bashrc:204]: Can we use `mktemp` here?

File: qa/workload-manifests/dnsutils.yaml [hunk 4]
```diff
@@ -0,0 +1,68 @@
...
+    imagePullPolicy: IfNotPresent
+    securityContext:
+       allowPrivilegeEscalation: false
+  restartPolicy: Always
+
+  ## https://kubernetes.io/docs/tasks/administer-cluster/dns-debugging-resolution/
+  ## kubectl exec -ti dnsutils -- cat /etc/resolv.conf
+  ## kubectl exec -i -t dnsutils -- nslookup kubernetes.default
+
+#security context example setting
```
Comment (briandowns) [inline qa/workload-manifests/dnsutils.yaml:47]: Do these commented lines need to be removed?

File: qa/qa-captain/bashrc [hunk 5]
```diff
@@ -0,0 +1,1116 @@
...
+        ;;
+    esac
+}
+
+# ======================================================================================
+# ================================== ~ ALIASES ~ =======================================
+# ======================================================================================
+#
+PATH=/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin:/usr/games:/usr/local/games:/snap/bin:/var/lib/rancher/k3s/bin:/var/lib/rancher/rke2/bin:/usr/local/bin/go/bin:/opt/rke2/bin/:${HOME}/.krew/bin
+alias ll="ls -lahr "
```
Comment (briandowns) [inline qa/qa-captain/bashrc:1087]: Looks like there are a couple different alias sections. Should these be consolidated into 1?

File: qa/qa-captain/bashrc [hunk 6]
```diff
@@ -0,0 +1,1116 @@
...
+
+# --- check for upgrade images in 
+
+
+# ======================================================================================
+# ================================== ~ GITHUB ~ =======================================
+# ======================================================================================
+
+# --- list issues in the github repos using the gh cli --- 
+ function ghls() {
```
Comment (briandowns) [inline qa/qa-captain/bashrc:1013]: Formatting seems to be messed up from here down.

File: qa/qa-captain/bashrc (hunk 1 above)
Comment (VestigeJ) [reply qa/qa-captain/bashrc:1]: I added this line to appease shellcheck really. So far none of the operating systems under test do not support Bash.

File: qa/qa-captain/bashrc (hunk 3 above)
Comment (VestigeJ) [reply qa/qa-captain/bashrc:174]: I've been leaving the config file in the user directory for QA to have an additional opportunity to edit the file before quickly launching a new installation re-using the previous config.yaml

File: qa/qa-captain/bashrc (hunk 5 above)
Comment (VestigeJ) [reply qa/qa-captain/bashrc:1057]: Yes - the alias' from the default Ubuntu bashrc were floating in the top. I've moved many down and removed what I imagine to be useless.

File: qa/workload-manifests/dnsutils.yaml (hunk 4 above)
Comment (VestigeJ) [reply qa/workload-manifests/dnsutils.yaml:47]: They're included links and example commands to ideally prevent someone from needing to leave their terminal in order to find additional information for how to use the tool.

File: qa/qa-captain/bashrc (hunk 6 above)
Comment (VestigeJ) [reply qa/qa-captain/bashrc:983]: You're correct - The number of functions within sections drop off and it's a bit of a junk drawer for functions to share with the team to add to their workstation shell environments. The sections CAPTAINS and GITHUB are both for that purpose.

File: qa/qa-captain/bashrc (hunk 2 above)
Comment (VestigeJ) [reply qa/qa-captain/bashrc]: I've removed this reference for a separate aliases file. The nodes aren't long lived enough to justify the house keeping of this in the bashrc.

File: qa/qa-captain/bashrc (hunk 1 above)
Comment (briandowns) [reply qa/qa-captain/bashrc:1]: The repo stanard is to use `/bin/sh`.

Comment (VestigeJ) [reply qa/qa-captain/bashrc:1]: This file is now inline with /bin/sh and shellcheck.

File: qa/qa-captain/bashrc [hunk 7]
```diff
@@ -0,0 +1,1083 @@
...
+# If this is an xterm set the title to user@host:dir
+case "${TERM}" in
+xterm*|rxvt*)
+    PS1="\[\e]0;${debian_chroot:+($debian_chroot)}\u@\h: \w\a\]$PS1"
+    ;;
+*)
+    ;;
+esac
+
+# enable programmable completion features
```
Comment (VestigeJ) [inline qa/qa-captain/bashrc:63]: This is the only included shellcheck ignore error/warning section. It's specifically for getting tab completion within /bin/bash as opposed to /bin/sh.

File: qa/qa-captain/bashrc [hunk 8]
```diff
@@ -0,0 +1,1083 @@
...
+    sudo /usr/local/bin/k3s-agent-killall.sh
+    sudo /usr/local/bin/k3s-agent-uninstall.sh
+    sudo /usr/local/bin/rke2-killall.sh
+    sudo /usr/local/bin/rke2-uninstall.sh
+    sudo /usr/local/bin/k3s-killall.sh
+    sudo /usr/local/bin/k3s-uninstall.sh
+    sudo /usr/bin/rke2-killall.sh
+    sudo /usr/bin/rke2-uninstall.sh
+    sudo /usr/bin/k3s-killall.sh
+    sudo /usr/bin/k3s-uninstall.sh
```
Comment (VestigeJ) [inline qa/qa-captain/bashrc:171]: This function is admittedly blunt - I'm rewriting this now to be more elegant.

**Review by briandowns — APPROVED** [review]

**Comment by briandowns** [comment]
After a second approval, we'll be good to SQUASH merge.

**Review by rancher-max — APPROVED** [review]
LGTM! Thanks for adding all of these!

**Review by brooksn — APPROVED** [review]

**Review by matttrach — APPROVED** [review]

File: qa/qa-captain/bashrc [hunk 9]
```diff
@@ -0,0 +1,1269 @@
...
+    echo "need to adjust this to detect which binaries are already installed - modify the path based on their presences"
+    echo "figure out all the steps here to setup a _rancher node to add downstream clusters to"
+    sudo apt update && sudo apt upgrade -y
+    wait
+    helm repo add rancher-latest https://releases.rancher.com/server-charts/latest
+    kubectl create namespace cattle-system
+    helm repo add jetstack https://charts.jetstack.io
+    helm repo update
+    wait
+    helm install cert-manager jetstack/cert-manager -n cert-manager --create-namespace --version v1.5.1
```
Comment (matttrach) [inline qa/qa-captain/bashrc:293]: does this work? how do you solve the dns queries to get certs?

File: qa/qa-captain/bashrc [hunk 10]
```diff
@@ -0,0 +1,1269 @@
...
+        "^rancher/hardened-ib-sriov-cni(.*)": "bcibase/hardened-ib-sriov-cni$1"
+        "^rancher/hardened-runc(.*)": "bcibase/hardened-runc$1"
+        "^rancher/hardened-rke2-cloud-provider(.*)": "bcibase/hardened-rke2-cloud-provider$1"
+        "^rancher/hardened-cni-plugins(.*)": "bcibase/hardened-cni-plugins$1"
+        "^rancher/hardened-dns-node-cache(.*)": "bcibase/hardened-dns-node-cache$1"
+        "^rancher/hardened-containerd(.*)": "bcibase/hardened-containerd$1"
+        "^rancher/hardened-cluster-autoscaler(.*)": "bcibase/hardened-cluster-autoscaler$1"
+        "^rancher/hardened-coredns(.*)": "bcibase/hardened-coredns$1"
+        "^rancher/hardened-multus-cni(.*)": "bcibase/hardened-multus-cni$1"
+        "^rancher/hardened-whereabouts(.*)": "bcibase/hardened-whereabouts$1"
```
Comment (matttrach) [inline qa/qa-captain/bashrc:450]: what is bcibase? Is this an internal mirror somewhere? It doesn't appear to be getting any updates on Dockerhub.

Comment (briandowns) [reply qa/qa-captain/bashrc:450]: I missed this in my initial review. There should not be any reference to the bcibase repositories. This is not to be used for anything. Great catch @matttrach

File: qa/qa-captain/bashrc (hunk 9 above)
Comment (rancher-max) [reply qa/qa-captain/bashrc:293]: Yep! Cert manager takes care of everything. This is directly from [rancher docs](https://ranchermanager.docs.rancher.com/pages-for-subheaders/install-upgrade-on-a-kubernetes-cluster#4-install-cert-manager) although we should probably update that cert manager version now. I've found `v1.8.0` works well in some previous testing.

File: qa/qa-captain/bashrc (hunk 10 above)
Comment (rancher-max) [reply qa/qa-captain/bashrc:450]: Yeah thanks for catching us out skimming 😬 @VestigeJ Can we update this to just setup a basic registries.yaml file now without the bcibase-specific testing stuff?

File: qa/qa-captain/bashrc (hunk 9 above)
Comment (VestigeJ) [reply qa/qa-captain/bashrc:290]: Updated to the later version Max referenced.

File: qa/qa-captain/bashrc (hunk 10 above)
Comment (VestigeJ) [reply qa/qa-captain/bashrc:444]: Pulled the function that checks for images built with bci

**Review by rancher-max — APPROVED** [review]

**Review by matttrach — APPROVED** [review]
Addressed all my concerns

**Pull request merged**